apiserver/encrypting-data-at-datastore-layer.md

```diff
@@ -0,0 +1,243 @@
+     // The values are regular expressions that correspond to the Golang regular expression language.
+     // +optional
+     AdditionalCORSAllowedOrigins []string `json:"additionalCORSAllowedOrigins,omitempty"`
++    // encryption allows the configuration of encryption of resources at the datastore layer.
++    // +optional
++    Encryption APIServerEncryption `json:"encryption"`
+ }
+
+ type APIServerServingCerts struct {
+@@ -63,6 +66,38 @@ type APIServerNamedServingCert struct {
+     ServingCertificate SecretNameReference `json:"servingCertificate"`
+ }
+
++type APIServerEncryption struct {
++    // type defines what encryption type should be used to encrypt resources at the datastore layer.
++    // When this field is unset (i.e. when it is set to the empty string), identity is implied.
++    // The behavior of unset can and will change over time.  Even if encryption is enabled by default,
++    // the meaning of unset may change to a different encryption type based on changes in best practices.
++    //
++    // When encryption is enabled, all sensitive resources shipped with the platform are encrypted.
++    // This list of sensitive resources can and will change over time.  The current authoritative list is:
++    //
++    //   1. secrets
++    //   2. config maps
++    //   3. routes
++    //   4. oauth access tokens
++    //   5. oauth authorize tokens
++    //
++    // +unionDiscriminator
++    // +optional
++    Type EncryptionType `json:"type,omitempty"`
++}
++
++type EncryptionType string
++
++const (
++    // identity refers to a type where no encryption is performed at the datastore layer.
++    // Resources are written as-is without encryption.
++    EncryptionTypeIdentity EncryptionType = "identity"
++
++    // aescbc refers to a type where AES-CBC with PKCS#7 padding and a 32-byte key
++    // is used to perform encryption at the datastore layer.
++    EncryptionTypeAESCBC EncryptionType = "aescbc"
++)
++
+ type APIServerStatus struct {
+ }
+```
+
+The default value of empty string for the type will imply `identity` (meaning no encryption of the datastore layer).
+
+The kube api server operator and openshift api server operator will observe this config and begin encrypting sensitive resources when it is set.  The operators are responsible for determining what resources to encrypt.
+
+### User Stories
+
+#### Story 1
+
+User enables encryption by setting `apiserver.spec.encryption.type` to `aescbc`.  After some time passes, user makes a backup of etcd.  The user confirms that the secret values are encrypted by checking to see if they have the `k8s:enc:aescbc:v1:` prefix.
+
+### Implementation Details/Notes/Constraints
+
+Key management is per resources (GroupResources). Different resources that are encrypted have different keys. Keys per resource are numbered with strict increasing unsigned integers. Different resources can have different or overlapping key numbers at the same time.
```

Comment:
I was expecting one key per server.  Doing multiple may cause more redeployments than expected

Reply:
I talked through this with @enj last week. My gut feeling is that one key per server is more natural. But cannot judge how many ripples that has through the controller logic.

We came to the conclusion that we have options to "lock step" rotations for different resources, i.e. do them all at once, even though each resource has its own key. That way we don't run into "endless" redeployments due to spread out rotations.